protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,515 @@
+            Ok(()) => Event::Removed(mapping),
+            Err(err) => Event::RemovalFailure(mapping, err),
+        })
+        .boxed()
+}
+
+/// A [`Provider::Gateway`] event.
+#[derive(Debug)]
+enum Event {
+    /// Port was successfully mapped.
+    Mapped(Mapping),
+    /// There was a failure mapping port.
+    MapFailure(Mapping, Box<dyn Error>),
+    /// Port was successfully removed.
+    Removed(Mapping),
+    /// There was a failure removing the mapping port.
+    RemovalFailure(Mapping, Box<dyn Error>),
+}
+
+/// Mapping of a Protocol and Port on the gateway.
+#[derive(Debug, Clone)]
+struct Mapping {
+    listener_id: ListenerId,
+    protocol: Protocol,
+    multiaddr: Multiaddr,
+    internal_addr: SocketAddrV4,
+}
+
+impl Hash for Mapping {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.listener_id.hash(state);
+    }
+}
+
+impl PartialEq for Mapping {
+    fn eq(&self, other: &Self) -> bool {
+        self.listener_id == other.listener_id
+    }
+}
+
+impl Eq for Mapping {}
+
+impl Borrow<ListenerId> for Mapping {
+    fn borrow(&self) -> &ListenerId {
+        &self.listener_id
+    }
+}
+
+/// Current state of a [`Mapping`].
+enum MappingState {
+    /// Port mapping is inactive, will be requested or re-requested on the next iteration.
+    Inactive,
+    /// Port mapping/removal has been requested on the gateway.
+    Pending,
+    /// Port mapping is active with the inner timeout.
+    Active(Delay),
+    /// Port mapping is permanent on the Gateway.
+    Permanent,
+}
+
+/// Current state of the UPnP [`Provider::Gateway`].
+enum GatewayState<P: Provider> {
+    Searching(BoxFuture<'static, Result<(P::Gateway, Ipv4Addr), Box<dyn std::error::Error>>>),
+    Available((Arc<P::Gateway>, Ipv4Addr)),
+    GatewayNotFound,
+}
+
+/// A `NetworkBehaviour` for UPnP port mapping. Automatically tries to map the external port
+/// to an internal address on the gateway on a `FromSwarm::NewListenAddr`.
+pub struct Behaviour<P>
+where
+    P: Provider,
+{
+    /// Gateway config.
+    config: Config,
+    /// UPnP interface state.
+    state: GatewayState<P>,
+
+    /// List of port mappings.
+    mappings: HashMap<Mapping, MappingState>,
+
+    /// Pending gateway events.
+    pending_events: FuturesUnordered<BoxFuture<'static, Event>>,
+}
+
+impl<P> Default for Behaviour<P>
+where
+    P: Provider + 'static,
+{
+    fn default() -> Self {
+        Self::new(Config::default())
+    }
+}
+
+impl<P> Behaviour<P>
+where
+    P: Provider + 'static,
+{
+    /// Builds a new `UPnP` behaviour.
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            state: GatewayState::Searching(P::search_gateway(config).boxed()),
+            mappings: Default::default(),
+            pending_events: Default::default(),
+        }
+    }
+}
+
+impl<P> NetworkBehaviour for Behaviour<P>
+where
+    P: Provider + 'static,
+{
+    type ConnectionHandler = dummy::ConnectionHandler;
+
+    type ToSwarm = ();
```

Comment:
Do we not want to emit that we successfully mapped / unmapped ports?

Reply:
Its a nice to have but i dont think its needed since `ToSwarm::ExternalAddrConfirmed` is enough to let other behaviours know about the external ip

Reply:
Its a nice to have but i dont think its needed since `ToSwarm::ExternalAddrConfirmed` is enough to let other behaviours know about the external ip

Reply:
It would just be purely informational. I am not 100% sure we need it but observing that upnp worked could be useful to the application.

Reply:
I am in favor of the additional events. Especially since we could then expose them as Prometheus metrics via [`libp2p-metrics`](https://github.com/libp2p/rust-libp2p/tree/master/misc/metrics). Not a blocker in my eyes though.

Reply:
yeah let's do it then :) is there a way to return two `ToSwarm` events from `poll` or should I just queue them?

Reply:
You would have to queue them. There is no way to return two in one `return`.